private-templates-service/adaptivecards-templating-service/src/storageproviders/MongoDBProvider.ts

```diff
@@ -0,0 +1,208 @@
+      .catch(e => {
+        return Promise.resolve({
+          success: false,
+          result: 0,
+          errorMessage: e
+        });
+      });
+  }
+  async updateTemplate(query: Partial<ITemplate>, updateQuery: Partial<ITemplate>): Promise<JSONResponse<Number>> {
+    return await this.Template.findOneAndUpdate(query, updateQuery)
+      .then(result => {
+        if (result) {
+          return Promise.resolve({ success: true, result: 1 });
+        }
+        return Promise.resolve({
+          success: false,
+          result: 0,
+          errorMessage: "No templates found matching given criteria."
+        });
+      })
+      .catch(e => {
+        return Promise.resolve({
+          success: false,
+          result: 0,
+          errorMessage: e
+        });
+      });
+  }
+  async insertUser(user: IUser): Promise<JSONResponse<Number>> {
+    return await this.User.create(user)
+      .then(result => {
+        return Promise.resolve({ success: true, result: 1 });
+      })
+      .catch(e => {
+        return Promise.resolve({
+          success: false,
+          number: 0,
+          errorMessage: e
+        });
+      });
+  }
+  async insertTemplate(template: ITemplate): Promise<JSONResponse<Number>> {
+    return await this.Template.create(template)
+      .then(result => {
+        return Promise.resolve({ success: true, result: 1 });
+      })
+      .catch(e => {
+        return Promise.resolve({
+          success: false,
+          number: 0,
+          errorMessage: e
+        });
+      });
+  }
+  async removeUser(query: Partial<IUser>): Promise<JSONResponse<Number>> {
+    return await this.User.deleteOne(query)
+      .then(result => {
+        if (result.deletedCount) {
+          return Promise.resolve({
+            success: true,
+            result: result.deletedCount
+          });
+        }
+        return Promise.resolve({
+          success: false,
+          result: 0,
+          errorMessage: "No users found matching given criteria"
+        });
+      })
+      .catch(e => {
+        return Promise.resolve({
+          success: false,
+          number: 0,
+          errorMessage: e
+        });
+      });
+  }
+  async removeTemplate(query: Partial<ITemplate>): Promise<JSONResponse<Number>> {
+    return await this.Template.deleteOne(query)
+      .then(result => {
+        if (result.deletedCount) {
+          return Promise.resolve({
+            success: true,
+            result: result.deletedCount
+          });
+        }
+        return Promise.resolve({
+          success: false,
+          result: 0,
+          errorMessage: "No templates found matching given criteria"
+        });
+      })
+      .catch(e => {
+        return Promise.resolve({
+          success: false,
+          number: 0,
+          errorMessage: e
+        });
+      });
+  }
+
+  // Set up mongoose models to work with the chosen collections
+  async setUpCollections(): Promise<JSONResponse<Boolean>> {
```

Comment:
I think this DB-specific stuff, from here, downwards - should be in a separate file.

Reply:
but all of this code is mongo specific - it's implementing **IStorageProvider** as common interface, why would it be in a different location?:)

Reply:
There are two separate things going on in this file - functions used for database _operations_, and then the database _setup/management_ methods. The IStorageProvider interface specifies the db operations, not the setup.

From looking quickly, the _constructor_/_initializeDB_/_connect_/_close_ could all be easily moved to a new file (or a separate class within the same file if that's easier).

Tagging @Jackbk for second opinion.

Reply:
I think @rachel-slater  has a good point here.  There is kinda two concepts going on. The MongoDBProvider can do adaptive card specific logic, e.g the `this.Template.findOneAndUpdate(query, updateQuery)` call. It should have a dependency on another layer which handles the grunt work.

Something like this stack:
AdaptiveCards Back end  -> MongoDBProvider -> MongoDBOperations

Anything not specific to the immediate running of adaptive cards (e.g, database set up) can be in operations.

It might seem a bit extra, but we have to remember that we are building an open source project and everything we commit from the start becomes an example for extensions later on.  If we can get a good pattern now, it will encourage any further DB providers to take a similar route.  And the easier we make it to understand the better/lower barrier to entry there will be.

Reply:
gotcha @rachel-slater , @Jackbk  - will do in a bit :)